src/status_im/ethereum/transactions/core.cljs

```diff
@@ -237,6 +236,30 @@
                   (conj (tx-history-end-reached checksum)))]
     (apply fx/merge cofx (tx-fetching-ended [checksum]) effects)))
 
+(fx/defn check-ens-transactions
+  [{:keys [db]} transfers]
+  (let [registrations (get db :ens/registrations)]
+    (doseq [[hash {:keys [state username custom-domain?]}] registrations]
+      (when (or (= state :dismissed) (= state :submitted))
+        (doseq [transaction transfers]
+          (let [transaction-hash (get transaction :hash)
+                type (get transaction :type)
+                transaction-success (get transaction :transfer)]
+            (when (= hash transaction-hash)
+              ; TODO Return from the loop once we find a match
+              (when (= transaction-success true)
+                (re-frame/dispatch [:update-ens-tx-state :success username custom-domain? hash])
```

Comment:
here are two things, events handlers MUST be pure, `re-frame/dispatch` is not pure, so you should use `:dispatch` fx, but we have internal rule to do not use `:dispath` , so you should use fx/merge here and instead `re-frame/dispatch` use `(update-ens-tx-state cofx ...)`

Reply:
We are getting there, nice work! This is not the easiest of the tasks to start with :)

So, now we basically have an effectful handler (btw any time you see `doseq` in clojure you know that is used for side effects).

We want to turn this into a pure function, as per what Mr Surfer said.

The way we do this in re-frame is instead of actually run the side-effects in the handler (i.e `(re-frame/dispatch [])`, we return a map with the side effects to run.

So for example if you want to make an HTTP request, you would return from the event-handler some like:

`{:db {....} :make-http-request [:param]}`

And then register an handler:

```
(re-frame/reg-fx :make-http-request
  (fn [params] ...))
```

https://github.com/day8/re-frame/blob/master/docs/Effects.md

In this case you are running side-effects only in order to `dispatch` other events, and run their event handlers.
As a general rule, we avoid having having event-handlers dispatching if they don't have to, and we instead accumulate the results of other event-handlers. To give you an example:

Say you have:

```
(fx/defn handle-a
  {:events [::event-a]}
  [cofx]
  (do-stuff-a))
```

and 

```
(fx/defn handle-b
 {:events [::event-b]}
  [cofx]
 (do-stuff-b))
```

Which reacts to`::event-a` and `::event-b`.

Then you have a new event `::event-c` which needs to run the same logic as both `::event-a` and `::event-b`.

One way you could do this (and this is what we did before) is to have something like this:


```
(fx/defn handle-c
  {:events [::event-c]}
 [cofx]
   {:db (modify-db-for-c cofx)
     :dispatch-n [[::event-a] [::event-b]]})
```

So essentially the handler of `::event-c` is further dispatching two events, similarly to the function you have above (the difference is that here is a pure function, and we run the side-effects in an handler which is provided by re-frame).

Eventually we moved to a different pattern, instead of further dispatching, we call the handlers directly and cumulate the changes, so we have something like:

```
(fx/defn handle-c
  {:events [::event-c]}
 [cofx]
  (fx/merge cofx
                  {:db (modify-db-for-c cofx)
                  (handle-a)
                  (handle-b)))
```

Which is equivalent (ish) in terms of code. This has some benefits, is more explict, easier to test, easier to follow the logic, and some drawbacks, handlers take longer to run (so perfomance might be an issue, as they don't yeild control for rendering),  they couple logic more heavily and adds complexity in accumlating them (`fx/merge` is our own macro to merge effects).
Regadless, most time for now we use this strategy.

So in your case, I would start by pulling all the objects you are interested in:
instead of:
```
 (let [registrations (get db :ens/registrations)]
    (doseq [[hash {:keys [state username custom-domain?]}] registrations]
      (when (or (= state :dismissed) (= state :submitted))
```

you can filter them, and already get those that match the transaction-hash in transfers:

```
 (let [set-of-transactions-hash (reduce (fn [acc {:keys [hash]}] (conj acc hash)) #{} transfers) 
         registrations (filter #(This function should match against the set, submitted and dismissed) (get db :ens/registrations)]
```
Which should give the "actionable" transactions (i.e those that match `transfers` and are `dismissed` or `submitted`.

The it's the tricky bit, you want to comulate the fxs.
How we do this is to use the form:

`(apply fx/merge cofx [vector of fxs])`

`apply` in clojure works similarly to `apply` in javascripts, minus the binding as that's not a thing in clojure, but they are effectively the same.

so what you want to do now, is to have a function that handles a single transaction and instead of dispatching calls the right handler:
```
(fx/defn whatever [cofx transaction] ;; This is an actionable transaction
  (cond 
     (:transaction-success transactions) ;; Is this two conditions mutually exclusive or a transaction (:transaction-success transactions) can be true an :failed ? 
     (fx/merge cofx (ens/update-ens-transaction-state..) (ens/save-username))
     (= :failed (:type transaction)
          (ens/update-ens-transacation-state ...))))
```

And finally you can comulate those and apply them, so roughly your code woudl be something like:


``` 
(let [set-of-transactions-hash (build-a-set-of-transactions-for-fast-loopkup) 
         registrations (filter-only-those-you-are-interested-in)
         fxs (map wathever registrations)]
        (apply fx/merge cofx fxs))
```

And that should do the tricky, there are other issues that might crop up, depending on how the code is structured. It would also be useful to add a few tests for this function, as it would be pure so you can easily make sure what you get in the db is what you expect.

This is very tricky stuff actually, very brave choice to start with! and sorry if I am too verbose or there's stuff you already know, let me know if you have any question or something does not make sense.

Reply:
Yeah i couldn't do it better than Mr. Cammellos, thanks

Reply:
@cammellos WOW! What a great explanation. Don't worry about being too verbose everything you said there is useful to me.

Also, it's not exactly my first issue tackled in react-status, but it's definitely the first that is that big haha.

Thanks again. I'm gonna fix it up!